py3status/modules/diskdata.py

```diff
@@ -165,12 +165,12 @@ def _get_io_stats(self, disk):
         write = 0
         with open('/proc/diskstats', 'r') as fd:
             for line in fd:
-                if disk and disk in line:
-                    data = line.split()
-                    read += int(data[5]) * self.sector_size
-                    write += int(data[9]) * self.sector_size
+                data = line.split()
+                if disk:
+                    if data[2] == disk:
+                        read += int(data[5]) * self.sector_size
+                        write += int(data[9]) * self.sector_size
                 else:
-                    data = line.split()
                     if data[1] == '0':
                         read += int(data[5]) * self.sector_size
```

Comment:
While you're here... See if this is correct.  Reason: I have 3 devices `sda`, `sr0`, and `dm-0` with minor device number `'0'`. This would be for `disk = None`... `# select None for all disks`. To me, it does not look like it would not read/write all (disk) lines, only those with `'0'`.

Also, It seems like we can add `break` too to prevent unnecessary repetitive looping + splitting after getting the `disk` values.

Reply:
Yep, you are right that as-is it will double-count when disks are not specifically selected on systems with some sort of logical volumes - dm-crypt/LUKS (dm-0 and the real device), and quite possibly something like LVM, as well. Unraveling this is difficult because /proc/diskstats does not provide enough information to know the arrangements of logical and physical volumes.

Reply:
> I don't know why we are checking '0'

I'll answer this first - the first and second columns are the device major and minor numbers. See Documentation/iostats.txt in the kernel source tree for more details on the format of /proc/diskstats.

> Would it be accurate if we remove if data[1] == '0': and count all lines for all disks?

Nope, because the contents of this file isn't just physical drives. Take the case where you have "sda" which has a partition "sda1" and then LUKS on top of that as "dm-0" containing the filesystem. Without the check for the minor device number of 0, any data written to that filesystem would be counted 3X! Note that with the check for the minor number, it's still counted 2X, so it's still wrong.

The problem here is that this data is not just physical disks. These are block devices - they can be physical, logical, or even something like a loopback device.

Reply:
Would this be a good solution?

To loop through all lines... Compute and put everything on the bar and offer `block_devices = []` config where users can filter out unwanted block devices. And to write down an explanation that this is `block_devices`, not physical disks... and that if we're using `dm-0`, `LVM`, etc... and that we may want to filter them out (noises).

This would be similar to using multiple `diskdata` with each disk specified. And to either depreciate (misleading) `all: ` or to make `all:` work with `block_devices = []` config for more accurate reading.

Reply:
I think that's a reasonable approach. The simple "disk" parameter is fully superseded and redundant via block_devices with this, which is kind of nice (and perhaps "disk" should be marked as a deprecated parameter so it could be removed at some point).